### ISSUE TITLE
Python 3 only?

### ISSUE THREAD

**Issue by ColdHeat** [CTFd/CTFd#655]
This is open for discussion. Should CTFd only support Python 3? I personally greatly prefer Python 2 but some features of Python 3 are nice.

**Comment by SkypLabs** [comment]
From the [official Python wiki](https://wiki.python.org/moin/Python2orPython3):

> Python 2.x is legacy, Python 3.x is the present and future of the language

Python 2 shouldn't be used at all for new projects.

**Comment by wroersma** [comment]
I mean you have to do it at some point might as well pull the trigger sooner than later right?

**Comment by ColdHeat** [comment]
I think the right decision is to be forwards compatible with Python 3. CTFd isn't a new project so I'm not sure what relevance the Python docs have here. Python 2 is legacy but more often used from my understanding. There is also not necessarily much benefit from switching to 3.0.

**Comment by anthraxx** [comment]
the lifetime of python2 is running out, i see no gain in supporting the past. just move on and accept py3 its not that bad :wink:

**Comment by SkypLabs** [comment]
> Python 2 is legacy but more often used from my understanding.

It's not because a lot of people use it that it means they are right to do so. Most of the time, programming languages are chosen and used because they are trendy, not because they are relevant for what people want to do with. For Python, Python 2 is still used today only because Python 3 is not backwards compatible with it and certainly not because of its qualities.

> There is also not necessarily much benefit from switching to 3.0.

I recommend you read this article: https://hackaday.com/2018/08/15/stop-using-python-2-what-you-need-to-know-about-python-3/

And it is just an overview of the advantages of Python 3 over Python 2. If you want to read more on the topic:

* https://medium.com/@GallegoDor/python-exploitation-1-input-ac10d3f4491f
* https://speakerdeck.com/pyconslides/python-3-dot-3-trust-me-its-better-than-python-2-dot-7-by-dr-brett-cannon
* https://igordavydenko.com/talks/by-pycon-2017

**Comment by ColdHeat** [comment]
Not every user uses Python 3 and not every user uses Docker. In addition, installing Python 3 and managing its dependencies is a massive pain. Switching to Python 3 would more than likely result in an uptick of users who need help dealing with Python and pip.

We are not being "legacy", we are being forwards-compatible. Once Python 2 is truly legacy there shouldn't be any problems with switching purely to Python 3. 

Unfortunately I never collected stats on what Python version is being used so I have no information to drive my decisions. Writing code for the sake of having modern code is in my opinion unnecessary. 

As the main contributor and supporter for CTFd, I just don't see the value. I think the strategy of supporting both until Python 2 is fully made legacy in 2020 is reasonable so I'm going to close this issue. Unless there is a great benefit specific to Python 3 (e.g. ASGI) Python 2 will be supported until 2020.

**Comment by SkypLabs** [comment]
Well, Python 2.7 has been released +8 years ago. It is already legacy, as said in the official Python documentation. Personally, I would have said that it's a piece of museum. It means that, by still using it, you prevent yourself and all of your contributors from using all of the new features and improvements brought by Python 3. The end user only manipulates the web interface of CTFd and doesn't care about which programming language is used by the back-end.

> Not every user uses Python 3 and not every user uses Docker.

And? I don't see where you go with that.

> In addition, installing Python 3 and managing its dependencies is a massive pain. Switching to Python 3 would more than likely result in an uptick of users who need help dealing with Python and pip.

I don't see why you suppose it is more difficult than with Python 2.

> We are not being "legacy", we are being forwards-compatible. Once Python 2 is truly legacy there shouldn't be any problems with switching purely to Python 3.

Which is great! I mean, it's better to be forwards-compatible than "legacy-only". It is not the topic of this discussion in my opinion.

> Unfortunately I never collected stats on what Python version is being used so I have no information to drive my decisions. Writing code for the sake of having modern code is in my opinion unnecessary.

Once again, it doesn't matter which Python version is being used the most since CTFd is a server-side application. It is only a matter for its developers.

> As the main contributor and supporter for CTFd, I just don't see the value. I think the strategy of supporting both until Python 2 is fully made legacy in 2020 is reasonable so I'm going to close this issue. Unless there is a great benefit specific to Python 3 (e.g. ASGI) Python 2 will be supported until 2020.

You opened this discussion to get the opinion of your contributors and I think you got our answer. I haven't seen any real argument in favour of Python 2.

**Comment by ColdHeat** [comment]
What new features and improvements are available in Python 3 that couldn't be adopted right now by a plugin or someone choosing to deploy on Python 3? 

It's quite simple, CTFd is an application that wants to increase its userbase as much as possible. Regardless of which Python version the user has, CTFd should be easy to run and install. Locking out every Python 2 user for very few gains would harm the adoption of the application. 

Because for many Python 2 is default. Its end of life is 2020 and Python 2 is currently getting some backported security fixes. That doesn't mean that other distros which are still supported are going to turn off Python 2 immediately.  Ubuntu 16.04, 14.04, 12.04 are all still supported at the moment and I don't think they install Python 3 by default. 

CTFd is a server side application that's deployed in places where there could be obtuse requirements, old distros, or running on Windows. Or the user could be non-technical. 

Installing and maintaining multiple Python versions is unquestionably a pain for a non-developer. I even struggle with it at times. 

There's a very simple argument for Python 2 support that I made earlier, only supporting Python 3 has no advantages for the project. The issue is titled "Python 3 only?" not is "Python 3 the more full-featured Python" or "Is Python 2 legacy?"

If moving entirely to Python 3 had a tangible user benefit I would more strongly consider being Python 3 *only* but I don't see anyone presenting any examples. 

Some examples that I would like but definitely not an exhaustive list:
 * Simpler Websockets instead of having to use SocketIO
 * Better ways of handling JSON inside of the database

Nothing in any of the data presented so far is a user-benefit that would help the entire ecosystem. As a CTFd user you are free to deploy on Python 3 and write your plugins in Python 3 only and get whatever Python 3 benefits you'd like. 

Put simply, there is no benefit to the project that would help users at large by restricting the only supported Python version to be Python 3. But there are users who would be locked out by by being Python 3 only.

**Comment by iptq** [comment]
Now that end of life is approaching and Python 3 is a lot more widespread, maybe reconsider this issue?

**Comment by ColdHeat** [comment]
@iptq See #820 

The point still stands that there isn't necessarily a good reason for CTFd to drop Python 2 support. After the EOL it will be re-evaluated.

**Comment by iptq** [comment]
Oh whoops, didn't see that

**Comment by ColdHeat** [comment]
If you have specific Python 3 concerns it will help if you air them in #820 to help push the discussion.
